File: cpp/cpp_100-999/449.cc
// C++
// leetcode 449
// https://leetcode.cn/problems/serialize-and-deserialize-bst/
// 序列化方法，每个数字范围都在[1,10000],可以用16bit表示，占两个字符
// 按先序遍历来序列化

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    int partOne = (1<<8) - 1;
    vector<char> vchar;
    
    void preOrder(TreeNode* root) {
        if(!root) return;
        int a = root->val;
        vchar.push_back(a>>8);
        vchar.push_back(a&partOne);
        preOrder(root->left);
        preOrder(root->right);
    }
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if(root == nullptr)
            return "";
        preOrder(root);
        string res(vchar.begin(), vchar.end());
        vchar.clear();
        return res;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if(data == "") return nullptr;
        int n = data.size();
        vector<int> tmp;
        for(int i = 0; i < n; i += 2) {
            int first = data[i];
            int second = (data[i+1] + 256 )% 256;
            int cur = (first<<8) + second;
            tmp.push_back(cur);
        }
        return inside(tmp, 0, tmp.size()-1);

    }

    TreeNode* inside(vector<int> &v, int l, int r) {
        int cur = v[l];
        TreeNode* ret = new TreeNode(cur);
        int mid = l + 1;
        while(mid <= r && v[mid] < cur)
            mid++;
        if(l < mid-1)
            ret->left = inside(v, l+1, mid-1);
        if(mid <= r)
            ret->right = inside(v, mid, r);
        return ret;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec* ser = new Codec();
// Codec* deser = new Codec();
// string tree = ser->serialize(root);
// TreeNode* ans = deser->deserialize(tree);
// return ans;

File: Go/Go_100-999/449.go
// Go
// leetcode 449
// https://leetcode.cn/problems/serialize-and-deserialize-bst/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import "strconv"

type Codec struct {
}

func Constructor() Codec {
	return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	return code(root)
}

func code(root *TreeNode) string {
	if root == nil {
		return "n"
	}
	num := strconv.Itoa(root.Val)
	return "(" + code(root.Left) + "!" + num + "!" + code(root.Right) + ")"
}

func decode(data string, l, r int) *TreeNode {
	if data[l] == 'n' {
		return nil
	}
	cnt := 0
	L := -1
	R := -1
	i := l
	for i = l; i <= r; i++ {
		switch data[i] {
		case '(':
			cnt++
		case ')':
			cnt--
		case '!':
			if cnt == 1 {
				if L < 0 {
					L = i
				} else {
					R = i
					break
				}
			}
		}
	}
	n, _ := strconv.Atoi(data[L+1 : R])
	var root *TreeNode = &TreeNode{n, nil, nil}
	root.Left = decode(data, l+1, L-1)
	root.Right = decode(data, R+1, r-1)
	return root
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	return decode(data, 0, len(data)-1)
}

/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor()
 * deser := Constructor()
 * tree := ser.serialize(root)
 * ans := deser.deserialize(tree)
 * return ans
 */